From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Mon, 17 Aug 2020 17:05:05 +0300
Subject: [PATCH 24/67] Revert "Bug 1333826 - Remove SDK_FILES, SDK_LIBRARY,
 and related is_sdk support in the build goop, r=mshal"

This reverts commit ca77995f5d2a21fba8191989b3ebfb913abc3eab.

Signed-off-by: Pavel Tumakaev <p.tumakaev@omprussia.ru>
---
 config/rules.mk                               | 14 ++++++++
 intl/unicharutil/util/moz.build               |  1 +
 mozglue/build/moz.build                       |  2 ++
 .../mozbuild/backend/recursivemake.py         | 14 +++++++-
 python/mozbuild/mozbuild/frontend/context.py  | 27 +++++++++++++++
 python/mozbuild/mozbuild/frontend/data.py     | 34 ++++++++++++++++---
 python/mozbuild/mozbuild/frontend/emitter.py  | 10 ++++++
 security/moz.build                            |  2 ++
 toolkit/library/moz.build                     |  2 ++
 xpcom/glue/standalone/moz.build               |  1 +
 xpcom/idl-parser/xpidl/moz.build              |  6 ++++
 11 files changed, 107 insertions(+), 6 deletions(-)

diff --git a/config/rules.mk b/config/rules.mk
index d03f5e7dbc5e..980ffae7b832 100644
--- a/config/rules.mk
+++ b/config/rules.mk
@@ -215,6 +215,8 @@ SIMPLE_PROGRAMS :=
 HOST_SHARED_LIBRARY :=
 HOST_PROGRAM :=
 HOST_SIMPLE_PROGRAMS :=
+SDK_LIBRARY :=
+WASM_LIBRARY :=
 endif
 
 ifdef MACH
@@ -836,6 +838,18 @@ ifneq (,$(DIST_SUBDIR)$(XPI_NAME))
 PREF_DIR = defaults/preferences
 endif
 
+################################################################################
+# SDK
+
+ifneq (,$(SDK_LIBRARY))
+ifndef NO_DIST_INSTALL
+SDK_LIBRARY_FILES := $(SDK_LIBRARY)
+SDK_LIBRARY_DEST := $(SDK_LIB_DIR)
+SDK_LIBRARY_TARGET := target
+INSTALL_TARGETS += SDK_LIBRARY
+endif
+endif # SDK_LIBRARY
+
 ################################################################################
 # CHROME PACKAGING
 
diff --git a/intl/unicharutil/util/moz.build b/intl/unicharutil/util/moz.build
index c93613429c91..5fc67b31f187 100644
--- a/intl/unicharutil/util/moz.build
+++ b/intl/unicharutil/util/moz.build
@@ -33,3 +33,4 @@ GeneratedFile(
 GeneratedFile("IsCombiningDiacritic.h", script="is_combining_diacritic.py")
 
 FINAL_LIBRARY = "xul"
+SDK_LIBRARY = True
diff --git a/mozglue/build/moz.build b/mozglue/build/moz.build
index 1fa4f8cba517..03b9e74e43c9 100644
--- a/mozglue/build/moz.build
+++ b/mozglue/build/moz.build
@@ -12,6 +12,8 @@ elif CONFIG["OS_TARGET"] in ("WINNT", "Darwin", "Android"):
 else:
     Library("mozglue")
 
+SDK_LIBRARY = True
+
 FORCE_STATIC_LIB = True
 STATIC_LIBRARY_NAME = "mozglue"
 
diff --git a/python/mozbuild/mozbuild/backend/recursivemake.py b/python/mozbuild/mozbuild/backend/recursivemake.py
index fa4d2b57b1cb..2e85660c441d 100644
--- a/python/mozbuild/mozbuild/backend/recursivemake.py
+++ b/python/mozbuild/mozbuild/backend/recursivemake.py
@@ -358,6 +358,7 @@ class RecursiveMakeBackend(MakeBackend):
         # used for a "magic" rm -rf.
         self._install_manifests["dist_public"]
         self._install_manifests["dist_private"]
+        self._install_manifests["dist_sdk"]
 
         self._traversal = RecursiveMakeTraversal()
         self._compile_graph = defaultdict(set)
@@ -1507,6 +1508,8 @@ class RecursiveMakeBackend(MakeBackend):
         backend_file.write("SHARED_LIBRARY := %s\n" % shared_lib)
         if libdef.soname:
             backend_file.write("DSO_SONAME := %s\n" % libdef.soname)
+        if libdef.is_sdk:
+            backend_file.write("SDK_LIBRARY := %s\n" % libdef.import_name)
         if libdef.symbols_file:
             if libdef.symbols_link_arg:
                 backend_file.write("EXTRA_DSO_LDOPTS += %s\n" % libdef.symbols_link_arg)
@@ -1519,6 +1522,8 @@ class RecursiveMakeBackend(MakeBackend):
         backend_file.write_once("LIBRARY_NAME := %s\n" % libdef.basename)
         backend_file.write("FORCE_STATIC_LIB := 1\n")
         backend_file.write("REAL_LIBRARY := %s\n" % libdef.lib_name)
+        if libdef.is_sdk:
+            backend_file.write("SDK_LIBRARY := %s\n" % libdef.import_name)
         if libdef.no_expand_lib:
             backend_file.write("NO_EXPAND_LIBS := 1\n")
 
@@ -1674,7 +1679,14 @@ class RecursiveMakeBackend(MakeBackend):
     def _process_final_target_files(self, obj, files, backend_file):
         target = obj.install_target
         path = mozpath.basedir(
-            target, ("dist/bin", "dist/xpi-stage", "_tests", "dist/include")
+            target,
+            (
+                "dist/bin",
+                "dist/xpi-stage",
+                "_tests",
+                "dist/include",
+                "dist/sdk",
+            ),
         )
         if not path:
             raise Exception("Cannot install to " + target)
diff --git a/python/mozbuild/mozbuild/frontend/context.py b/python/mozbuild/mozbuild/frontend/context.py
index c7ca0e4e7bfe..261bf910acb2 100644
--- a/python/mozbuild/mozbuild/frontend/context.py
+++ b/python/mozbuild/mozbuild/frontend/context.py
@@ -2027,6 +2027,33 @@ VARIABLES = {
         them correctly.
         """,
     ),
+
+    "SDK_FILES": (
+        ContextDerivedTypedHierarchicalStringList(Path),
+        list,
+        """List of files to be installed into the sdk directory.
+
+        ``SDK_FILES`` will copy (or symlink, if the platform supports it)
+        the contents of its files to the ``dist/sdk`` directory. Files that
+        are destined for a subdirectory can be specified by accessing a field.
+        For example, to export ``foo.py`` to the top-level directory and
+        ``bar.py`` to the directory ``subdir``, append to
+        ``SDK_FILES`` like so::
+
+           SDK_FILES += ['foo.py']
+           SDK_FILES.subdir += ['bar.py']
+        """,
+    ),
+
+    "SDK_LIBRARY": (
+        bool,
+        bool,
+        """Whether the library built in the directory is part of the SDK.
+
+        The library will be copied into ``SDK_LIB_DIR`` (``$DIST/sdk/lib``).
+        """,
+    ),
+
     "SIMPLE_PROGRAMS": (
         StrictOrderingOnAppendList,
         list,
diff --git a/python/mozbuild/mozbuild/frontend/data.py b/python/mozbuild/mozbuild/frontend/data.py
index cb0492f67c97..3e780a60fa21 100644
--- a/python/mozbuild/mozbuild/frontend/data.py
+++ b/python/mozbuild/mozbuild/frontend/data.py
@@ -668,11 +668,14 @@ class Library(BaseLibrary):
     """Context derived container object for a library"""
 
     KIND = "target"
-    __slots__ = ()
+    __slots__ = (
+        "is_sdk",
+    )
 
-    def __init__(self, context, basename, real_name=None):
+    def __init__(self, context, basename, real_name=None, is_sdk=False):
         BaseLibrary.__init__(self, context, real_name or basename)
         self.basename = basename
+        self.is_sdk = is_sdk
 
 
 class StaticLibrary(Library):
@@ -681,9 +684,15 @@ class StaticLibrary(Library):
     __slots__ = ("link_into", "no_expand_lib")
 
     def __init__(
-        self, context, basename, real_name=None, link_into=None, no_expand_lib=False
+        self,
+        context,
+        basename,
+        real_name=None,
+        is_sdk=False,
+        link_into=None,
+        no_expand_lib=False,
     ):
-        Library.__init__(self, context, basename, real_name)
+        Library.__init__(self, context, basename, real_name, is_sdk)
         self.link_into = link_into
         self.no_expand_lib = no_expand_lib
 
@@ -841,12 +850,13 @@ class SharedLibrary(Library):
         context,
         basename,
         real_name=None,
+        is_sdk=False,
         soname=None,
         variant=None,
         symbols_file=False,
     ):
         assert variant in range(1, self.MAX_VARIANT) or variant is None
-        Library.__init__(self, context, basename, real_name)
+        Library.__init__(self, context, basename, real_name, is_sdk)
         self.variant = variant
         self.lib_name = real_name or basename
         self.output_category = context.get("SHARED_LIBRARY_OUTPUT_CATEGORY")
@@ -1354,6 +1364,20 @@ class Exports(FinalTargetFiles):
         return "dist/include"
 
 
+class SdkFiles(FinalTargetFiles):
+    """Sandbox container object for SDK_FILES, which is a
+    HierarchicalStringList.
+
+    We need an object derived from ContextDerived for use in the backend, so
+    this object fills that role. It just has a reference to the underlying
+    HierarchicalStringList, which is created when parsing SDK_FILES.
+    """
+
+    @property
+    def install_target(self):
+        return "dist/sdk"
+
+
 class GeneratedFile(ContextDerived):
     """Represents a generated file."""
 
diff --git a/python/mozbuild/mozbuild/frontend/emitter.py b/python/mozbuild/mozbuild/frontend/emitter.py
index 0e763ea1d6f0..a3bcd9528e04 100644
--- a/python/mozbuild/mozbuild/frontend/emitter.py
+++ b/python/mozbuild/mozbuild/frontend/emitter.py
@@ -61,6 +61,7 @@ from .data import (
     RustProgram,
     RustTests,
     SandboxedWasmLibrary,
+    SdkFiles,
     SharedLibrary,
     SimpleProgram,
     Sources,
@@ -846,6 +847,14 @@ class TreeMetadataEmitter(LoggingMixin):
                     )
                 shared_args["soname"] = soname
 
+            # If both a shared and a static library are created, only the
+            # shared library is meant to be a SDK library.
+            if context.get("SDK_LIBRARY"):
+                if shared_lib:
+                    shared_args["is_sdk"] = True
+                elif static_lib:
+                    static_args["is_sdk"] = True
+
             if context.get("NO_EXPAND_LIBS"):
                 if not static_lib:
                     raise SandboxValidationError(
@@ -1509,6 +1518,7 @@ class TreeMetadataEmitter(LoggingMixin):
             ("MOZ_SRC_FILES", MozSrcFiles),
             ("OBJDIR_FILES", ObjdirFiles),
             ("OBJDIR_PP_FILES", ObjdirPreprocessedFiles),
+            ("SDK_FILES", SdkFiles),
             ("TEST_HARNESS_FILES", TestHarnessFiles),
         ):
             if var == "MOZ_SRC_FILES":
diff --git a/security/moz.build b/security/moz.build
index 90bae217e5c8..c4a03e14cfed 100644
--- a/security/moz.build
+++ b/security/moz.build
@@ -33,6 +33,8 @@ if CONFIG["MOZ_FOLD_LIBS"]:
     # TODO: The library name can be changed when bug 845217 is fixed.
     SHARED_LIBRARY_NAME = "nss3"
 
+    SDK_LIBRARY = True
+
     USE_LIBS += [
         "nspr4",
         "nss3_static",
diff --git a/toolkit/library/moz.build b/toolkit/library/moz.build
index 2e194a1abaf7..936b5416f31a 100644
--- a/toolkit/library/moz.build
+++ b/toolkit/library/moz.build
@@ -113,6 +113,8 @@ def Libxul(name, output_category=None):
 # is then linked to both build/libxul and gtest/libxul.
 Library("xul")
 
+SDK_LIBRARY = True
+
 STATIC_LIBRARY_NAME = "xul_s"
 
 # Use a FINAL_LIBRARY for Libxul_defines to propagate from there. They wouldn't
diff --git a/xpcom/glue/standalone/moz.build b/xpcom/glue/standalone/moz.build
index b933b8daa0b7..91495db6135e 100644
--- a/xpcom/glue/standalone/moz.build
+++ b/xpcom/glue/standalone/moz.build
@@ -10,6 +10,7 @@ SOURCES += [
 
 Library("xpcomglue")
 
+SDK_LIBRARY = True
 FORCE_STATIC_LIB = True
 STATIC_LIBRARY_NAME = "xpcomglue"
 
diff --git a/xpcom/idl-parser/xpidl/moz.build b/xpcom/idl-parser/xpidl/moz.build
index 06cdd3330718..d3c3e6eb3f51 100644
--- a/xpcom/idl-parser/xpidl/moz.build
+++ b/xpcom/idl-parser/xpidl/moz.build
@@ -7,3 +7,9 @@ PYTHON_UNITTEST_MANIFESTS += [
 ]
 
 GeneratedFile("xpidl.stub", script="header.py", entry_point="main")
+
+SDK_FILES.bin += [
+    "header.py",
+    "jsonxpt.py",
+    "xpidl.py",
+]
